src/libguac/user-handlers.h

```diff
@@ -186,4 +220,73 @@ __guac_instruction_handler __guac_handle_disconnect;
  */
 extern __guac_instruction_handler_mapping __guac_instruction_handler_map[];
 
+/**
+ * Handler mapping table for instructions (opcodes) specifically for the
+ * handshake portion of the connection.  Each
+ * __guac_instruction_handler_mapping structure within this NULL-terminated
+ * array maps an opcode to a __guac_instruction_handler.  The end of the array
+ * must be marked with a mapping with the opcode set to NULL.
+ */
+extern __guac_instruction_handler_mapping __guac_handshake_handler_map[];
+
+/**
+ * Frees the given array of mimetypes, including the space allocated to each
+ * mimetype string within the array. The provided array of mimetypes MUST have
+ * been allocated with guac_copy_mimetypes().
+ *
+ * @param mimetypes
+ *     The NULL-terminated array of mimetypes to free. This array MUST have
+ *     been previously allocated with guac_copy_mimetypes().
+ */
+void guac_free_mimetypes(char** mimetypes);
+
+/**
+ * Copies the given array of mimetypes (strings) into a newly-allocated NULL-
+ * terminated array of strings. Both the array and the strings within the array
+ * are newly-allocated and must be later freed via guac_free_mimetypes().
+ *
+ * @param mimetypes
+ *     The array of mimetypes to copy.
+ *
+ * @param count
+ *     The number of mimetypes in the given array.
+ *
+ * @return
+ *     A newly-allocated, NULL-terminated array containing newly-allocated
+ *     copies of each of the mimetypes provided in the original mimetypes
+ *     array.
+ */
+char** guac_copy_mimetypes(char** mimetypes, int count);
+
+/**
+ * Call the appropriate handler defined by the given user for the given
+ * instruction. A comparison is made between the instruction opcode and the
+ * initial handler lookup table defined in the map that is provided to this
+ * function. If an entry for the instruction is found in the provided map,
+ * the handler defined in that map will be called and the value returned.  If
+ * no match is found, it is silently ignored.
+ *
+ * @param map
+ *     The array that holds the opcode to handler mappings.
+ * 
+ * @param user
+ *     The user whose handlers should be called.
+ *
+ * @param opcode
+ *     The opcode of the instruction to pass to the user via the appropriate
+ *     handler.
+ *
+ * @param argc
+ *     The number of arguments which are part of the instruction.
+ *
+ * @param argv
+ *     An array of all arguments which are part of the instruction.
+ *
+ * @return
+ *     Non-negative if the instruction was handled successfully, or negative
```

Comment:
In at least one case, the return value of this function is presumed to be successful only if zero, with any non-zero value indicating an error:

https://github.com/apache/guacamole-server/blob/98cb7ccf6773f9e55645ccfe542de31ab2bc136c/src/libguac/user-handshake.c#L273-L275

The above usage matches with the underlying definition of `__guac_instruction_handler`:

https://github.com/apache/guacamole-server/blob/98cb7ccf6773f9e55645ccfe542de31ab2bc136c/src/libguac/user-handlers.h#L52-L55

As `__guac_user_call_opcode_handler()` passes through the return value of the matching `__guac_instruction_handler`, this should also be zero if successful, non-zero otherwise.

Reply:
Adjusted comment/docs to match `__guac_instruction_handler()`